File: Data Structures & Algorithms/add-two-numbers/submission-6.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {

        // carry = 0
        int carry = 0;
        // sum = 0
        int sum = 0;
        // result
        ListNode result = new ListNode(0);
        ListNode first = result;

        // while l1 and l2 not null 
           while (true){
        //      if l1 not null 
                // if (l1 == null){
                //     break;
                // }
                if (l1 != null){
        //          sum += l1
                    sum = sum + l1.val;
        //          go to next node; 
                    if (l1 != null){
                        l1 = l1.next;
                    }
                }
        //      if l2 not null
                if (l2 != null){
        //          sum += l2
                    sum = sum + l2.val;
        //          got to next node
                    l2 = l2.next;
                    
                }
        //      sum += carry
                sum = sum + carry;
                //carry = 0;
        //      carry += carry(sum)
                carry = sum/10;
        //      put non carry part of sum in result
                result.val = sum%10;
        //      add node to result and go to it
                if (l1 == null && l2 == null){
                    break;
                }
                result.next = new ListNode(0);
                result = result.next;
        //      sum = 0 
                sum = 0;
           }

        // add carry to end of result 
        if(carry > 0){
            if(carry >= 10){
                result.next = new ListNode(carry%10);
                result = result.next;
                result.next = new ListNode(carry%10);
            }
            else{
                result.next = new ListNode(carry);
            }
        }
        return first;
        
    }
}

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-6.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:

        table = {
            (None, None): lambda n1,n2 : None,
            (0, None): lambda n1,n2: n1,
            (None, 0): lambda n1,n2: n2,
            (0, 0): lambda n1,n2: n1 if n1.val <= n2.val else n2
        }

        def incr(n):
            nonlocal list1
            nonlocal list2
            if list1 == n:
                list1 = list1.next
            else:
                list2 = list2.next
        
        prev = root = table[(0 if list1 else None, 0 if list2 else None)](list1, list2)
        if not root:
            return None
        incr(root)
        while table[(0 if list1 else None, 0 if list2 else None)](list1, list2):
            curr = table[(0 if list1 else None, 0 if list2 else None)](list1, list2)
            incr(curr)
            prev.next = curr
            prev = curr
        prev.next = None
        return root

        
        


                